geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/AbstractSimpleFeature.scala

```diff
@@ -0,0 +1,223 @@
+
+    def setAttributeNoConvert(index: Int, value: AnyRef): Unit
+    def setId(id: String): Unit = featureId.setID(id)
+
+    override def getIdentifier: FeatureId = featureId
+    override def getID: String = featureId.getID // this needs to reference the featureId, as it can be updated
+    override def getUserData: java.util.Map[AnyRef, AnyRef] = userData
+
+    override def setAttribute(name: Name, value: Object): Unit = setAttribute(name.getLocalPart, value)
+    override def setAttribute(name: String, value: Object): Unit = {
+      val index = sft.indexOf(name)
+      if (index == -1) {
+        throw new IllegalArgumentException(s"Attribute $name does not exist in type $sft")
+      }
+      setAttribute(index, value)
+    }
+    override def setAttribute(index: Int, value: Object): Unit = {
+      val binding = sft.getDescriptor(index).getType.getBinding
+      setAttributeNoConvert(index, Converters.convert(value, binding).asInstanceOf[AnyRef])
+    }
+
+    // following methods delegate to setAttribute to get type conversion
+    override def setAttributes(vals: java.util.List[Object]): Unit = {
+      var i = 0
+      while (i < vals.size) {
+        setAttribute(i, vals.get(i))
+        i += 1
+      }
+    }
+    override def setAttributes(vals: Array[Object]): Unit = {
+      var i = 0
+      while (i < vals.length) {
+        setAttribute(i, vals(i))
+        i += 1
+      }
+    }
+    override def setDefaultGeometry(geo: Object): Unit = setAttribute(geomIndex, geo)
+    override def setValue(newValue: Object): Unit = setValue(newValue.asInstanceOf[java.util.Collection[Property]])
+    override def setValue(values: java.util.Collection[Property]): Unit = {
+      import scala.collection.JavaConversions._
+      var i = 0
+      values.foreach { p =>
+        setAttribute(i, p.getValue)
+        i += 1
+      }
+    }
+    override def setDefaultGeometryProperty(geoAttr: GeometryAttribute): Unit =
+      if (geoAttr == null) { setDefaultGeometry(null) } else { setDefaultGeometry(geoAttr.getValue) }
+  }
+}
+
+/**
+  * Base class for simple feature implementations, with boilerplate
+  *
+  * @param sft simple feature type
+  */
+abstract class AbstractSimpleFeature(sft: SimpleFeatureType) extends SimpleFeature {
+
+  lazy protected val geomDesc  = sft.getGeometryDescriptor
+  lazy protected val geomIndex = if (geomDesc == null) { -1 } else { sft.indexOf(geomDesc.getLocalName) }
+
+  override def getFeatureType: SimpleFeatureType = sft
+  override def getType: SimpleFeatureType = sft
+  override def getName: Name = sft.getName
+
+  override def getAttribute(name: Name): AnyRef = getAttribute(name.getLocalPart)
+  override def getAttribute(name: String): AnyRef = {
+    val index = sft.indexOf(name)
+    if (index == -1) { null } else { getAttribute(index) }
+  }
+
+  override def getAttributeCount: Int = sft.getAttributeCount
+  override def getAttributes: java.util.List[Object] = {
+    val list = new java.util.ArrayList[AnyRef]
```

Comment:
I kinda assume you leave a few quickies like this to see if I actually read your PRs....